nuxeo-tree-snapshot-core/src/main/java/org/nuxeo/snapshot/SnapshotableAdapter.java

```diff
@@ -132,19 +137,38 @@ public Snapshot createSnapshot(VersioningOption option) {
 
         DocumentModelList children = doc.getCoreSession().getChildren(doc.getRef());
 
-        String[] vuuids = new String[children.size()];
-
-        for (int i = 0; i < children.size(); i++) {
-            DocumentModel child = children.get(i);
-            if (!child.isFolder()) {
-                DocumentRef leafRef = createLeafVersion(child, option);
-                vuuids[i] = leafRef.toString();
+        List<DocumentModel> folders = new ArrayList<>();
+        List<DocumentModel> leafs = new ArrayList<>();
+        doc.getCoreSession().save();
+        IterableQueryResult res = doc.getCoreSession().queryAndFetch(
+                "SELECT ecm:uuid FROM Document WHERE ecm:parentId = '" + doc.getId() + "'",
+                "NXQL");
+
+        String[] vuuids = new String[(int) res.size()];
+        for (Map<String, Serializable> item : res) {
+            DocumentModel child = doc.getCoreSession().getDocument(new IdRef((String) item.get(NXQL.ECM_UUID)));
+            if (child.isFolder()) {
+                folders.add(child);
             } else {
-                SnapshotableAdapter adapter = new SnapshotableAdapter(child);
-                Snapshot snap = adapter.createSnapshot(option);
-                vuuids[i] = snap.getRef().toString();
+                leafs.add(child);
             }
         }
+        res.close();
```

Comment:
Do the `close` automatically and safely by allocating the `IterableQueryResult` in a try-with-resources block.